File: src/ios/AmiraWellness/AmiraWellness/UI/Screens/Progress/EmotionalTrendsViewModel.swift
import Foundation // For basic data types and Date operations
import Combine // For Combine framework features
import SwiftUI // For ObservableObject and Published property wrappers

// Internal imports for data models, services, and utilities
import EmotionalTrend // Core data model for emotional trends
import EmotionalTrendResponse // Response data structure from trend analysis
import EmotionalInsight // Insights derived from emotional trend analysis
import TrendPeriodType // Time periods for trend analysis
import EmotionType // Emotion types for filtering and display
import ProgressService // Service for accessing progress and trend data
import Logger // Logging service for debugging and error tracking
import Haptics // Haptic feedback for user interactions

/// ViewModel for the Emotional Trends screen that manages the state and business logic for visualizing emotional patterns over time
@MainActor
@ObservableObject
class EmotionalTrendsViewModel {
    
    // MARK: - Private Properties
    
    private let progressService: ProgressService // Service for fetching emotional trends
    
    // MARK: - Published Properties
    
    @Published var trends: [EmotionalTrend] = [] // Array of emotional trends to display
    @Published var insights: [EmotionalInsight] = [] // Array of insights related to the trends
    @Published var selectedPeriod: TrendPeriodType = .weekly // Currently selected time period for trend analysis
    @Published var selectedEmotionTypes: [EmotionType] = [] // Array of selected emotion types for filtering
    @Published var isLoading: Bool = false // Indicates if the data is currently being loaded
    @Published var errorMessage: String? // Error message to display if loading fails
    @Published var selectedTrend: EmotionalTrend? // The currently selected trend for detailed viewing
    @Published var startDate: Date // Start date for the current trend period
    @Published var endDate: Date // End date for the current trend period
    
    private var cancellables: Set<AnyCancellable> = [] // Set to hold Combine cancellables
    
    // MARK: - Initialization
    
    /// Initializes the EmotionalTrendsViewModel with dependencies
    /// - Parameter progressService: Optional ProgressService for fetching emotional trends
    init(progressService: ProgressService? = nil) {
        // Store the provided progressService or create a new instance
        self.progressService = progressService ?? ProgressService()
        // Initialize trends as an empty array
        self.trends = []
        // Initialize insights as an empty array
        self.insights = []
        // Set selectedPeriod to .weekly by default
        self.selectedPeriod = .weekly
        // Initialize selectedEmotionTypes as an empty array (all emotions)
        self.selectedEmotionTypes = []
        // Set isLoading to false
        self.isLoading = false
        // Set errorMessage to nil
        self.errorMessage = nil
        // Set selectedTrend to nil
        self.selectedTrend = nil
        // Initialize date range from selectedPeriod.defaultRange()
        (self.startDate, self.endDate) = selectedPeriod.defaultRange()
        // Initialize cancellables set for storing subscriptions
        self.cancellables = []
    }
    
    // MARK: - Public Methods
    
    /// Loads emotional trend data for the selected period and emotion types
    func loadTrends() {
        // Set isLoading to true
        isLoading = true
        // Clear any previous error message
        errorMessage = nil
        
        // Call progressService.getEmotionalTrends with selectedPeriod
        progressService.getEmotionalTrends(periodType: selectedPeriod, startDate: startDate, endDate: endDate) { [weak self] result in
            guard let self = self else { return }
            
            switch result {
            case .success(let trendResponse):
                // Update trends and insights properties
                self.trends = trendResponse.trends
                self.insights = trendResponse.insights
                
                // Filter trends by selectedEmotionTypes if not empty
                if !self.selectedEmotionTypes.isEmpty {
                    self.applyEmotionTypeFilters()
                }
                
                // Sort trends by averageIntensity or occurrence count
                self.trends.sort { $0.averageIntensity > $1.averageIntensity }
                
                // Set isLoading to false
                self.isLoading = false
                
            case .failure(let error):
                // Set errorMessage with appropriate error
                self.errorMessage = "Failed to load trends: \(error)"
                // Set isLoading to false
                self.isLoading = false
            }
        }
    }
    
    /// Loads emotional trend data using async/await
    @available(iOS 15.0, *)
    func loadTrendsAsync() async {
        // Set isLoading to true
        isLoading = true
        // Clear any previous error message
        errorMessage = nil
        
        do {
            // Try to call progressService.getEmotionalTrendsAsync with selectedPeriod
            let trendResponse = try await progressService.getEmotionalTrendsAsync(periodType: selectedPeriod)
            
            // Update trends and insights properties
            self.trends = trendResponse.trends
            self.insights = trendResponse.insights
            
            // Filter trends by selectedEmotionTypes if not empty
            if !self.selectedEmotionTypes.isEmpty {
                self.applyEmotionTypeFilters()
            }
            
            // Sort trends by averageIntensity or occurrence count
            self.trends.sort { $0.averageIntensity > $1.averageIntensity }
            
            // Set isLoading to false
            self.isLoading = false
        } catch {
            // Set errorMessage with appropriate error
            self.errorMessage = "Failed to load trends: \(error)"
            // Set isLoading to false
            self.isLoading = false
        }
    }
    
    /// Changes the selected time period for trend analysis
    /// - Parameter periodType: The new time period to select
    func changePeriod(periodType: TrendPeriodType) {
        // Set selectedPeriod to the provided periodType
        selectedPeriod = periodType
        // Update startDate and endDate from periodType.defaultRange()
        (startDate, endDate) = periodType.defaultRange()
        // Call loadTrends() to refresh data with new period
        loadTrends()
        // Provide haptic feedback for the change
        Haptics.shared.generateFeedback(.selection)
    }
    
    /// Toggles the inclusion of an emotion type in the filter
    /// - Parameter emotionType: The emotion type to toggle
    func toggleEmotionTypeFilter(emotionType: EmotionType) {
        // Check if emotionType is already in selectedEmotionTypes
        if selectedEmotionTypes.contains(emotionType) {
            // If present, remove it from the array
            selectedEmotionTypes.removeAll { $0 == emotionType }
        } else {
            // If not present, add it to the array
            selectedEmotionTypes.append(emotionType)
        }
        
        // Apply the filter to the current trends
        applyEmotionTypeFilters()
        // Provide haptic feedback for the change
        Haptics.shared.generateFeedback(.selection)
    }
    
    /// Clears all emotion type filters
    func clearEmotionTypeFilters() {
        // Set selectedEmotionTypes to an empty array
        selectedEmotionTypes = []
        // Reload trends with no emotion type filter
        loadTrends()
        // Provide haptic feedback for the change
        Haptics.shared.generateFeedback(.selection)
    }
    
    /// Selects a trend for detailed viewing
    /// - Parameter trend: The trend to select
    func selectTrend(trend: EmotionalTrend?) {
        // Set selectedTrend to the provided trend
        selectedTrend = trend
        // If trend is not nil, provide haptic feedback for selection
        if trend != nil {
            Haptics.shared.generateFeedback(.light)
        }
    }
    
    /// Filters insights related to a specific emotion type
    /// - Parameter emotionType: The emotion type to filter by
    /// - Returns: Array of insights related to the emotion type
    func getInsightsForEmotion(emotionType: EmotionType) -> [EmotionalInsight] {
        // Filter insights array to include only those related to the specified emotionType
        let filteredInsights = insights.filter { $0.relatedEmotions.contains(emotionType) }
        // Sort filtered insights by confidence level (descending)
        return filteredInsights.sorted { $0.confidence > $1.confidence }
    }
    
    /// Returns the top insights based on confidence level
    /// - Parameter limit: The maximum number of insights to return
    /// - Returns: Array of top insights
    func getTopInsights(limit: Int) -> [EmotionalInsight] {
        // Sort insights array by confidence level (descending)
        let sortedInsights = insights.sorted { $0.confidence > $1.confidence }
        // Return the first 'limit' number of insights
        return Array(sortedInsights.prefix(limit))
    }
    
    /// Updates the date range for trend analysis
    /// - Parameters:
    ///   - start: The new start date
    ///   - end: The new end date
    func updateDateRange(start: Date, end: Date) {
        // Validate that start date is before end date
        guard start < end else {
            return
        }
        
        // Set startDate to the provided start date
        startDate = start
        // Set endDate to the provided end date
        endDate = end
        // Call loadTrends() to refresh data with new date range
        loadTrends()
    }
    
    /// Determines the most frequently occurring emotion in the trends
    /// - Returns: The most frequent emotion type or nil if no trends
    func getMostFrequentEmotion() -> EmotionType? {
        // Check if trends array is empty, return nil if so
        guard !trends.isEmpty else {
            return nil
        }
        
        // Group trends by emotionType and count occurrences
        let emotionCounts = Dictionary(grouping: trends, by: { $0.emotionType })
            .mapValues { $0.count }
        
        // Find the emotion type with the highest count
        let mostFrequentEmotion = emotionCounts.max { a, b in a.value < b.value }?.key
        
        // Return the most frequent emotion type
        return mostFrequentEmotion
    }
    
    /// Calculates the average intensity for a specific emotion type
    /// - Parameter emotionType: The emotion type to calculate the average intensity for
    /// - Returns: Average intensity or 0 if no data
    func getAverageIntensityForEmotion(emotionType: EmotionType) -> Double {
        // Filter trends to include only those matching the specified emotionType
        let filteredTrends = trends.filter { $0.emotionType == emotionType }
        
        // If no matching trends, return 0
        guard !filteredTrends.isEmpty else {
            return 0
        }
        
        // Calculate the average of averageIntensity values from matching trends
        let totalIntensity = filteredTrends.reduce(0.0) { $0 + $1.averageIntensity }
        let averageIntensity = totalIntensity / Double(filteredTrends.count)
        
        // Return the calculated average
        return averageIntensity
    }
    
    // MARK: - Private Methods
    
    /// Applies the current emotion type filters to the trends
    private func applyEmotionTypeFilters() {
        // Check if selectedEmotionTypes is empty (no filtering)
        if selectedEmotionTypes.isEmpty {
            // If empty, reload all trends from the original data
            loadTrends()
        } else {
            // If not empty, filter trends to include only those with emotionType in selectedEmotionTypes
            trends = trends.filter { selectedEmotionTypes.contains($0.emotionType) }
        }
    }
}

File: src/android/app/src/main/kotlin/com/amirawellness/ui/screens/progress/EmotionalTrendsViewModel.kt
package com.amirawellness.ui.screens.progress

import androidx.lifecycle.ViewModel // androidx.lifecycle:lifecycle-viewmodel:2.6.1
import androidx.lifecycle.viewModelScope // androidx.lifecycle:lifecycle-viewmodel-ktx:2.6.1
import com.amirawellness.domain.usecases.emotional.GetEmotionalTrendsUseCase // src/android/app/src/main/kotlin/com/amirawellness/domain/usecases/emotional/GetEmotionalTrendsUseCase.kt
import com.amirawellness.data.models.EmotionalTrend // src/android/app/src/main/kotlin/com/amirawellness/data/models/EmotionalTrend.kt
import com.amirawellness.data.models.EmotionalInsight // src/android/app/src/main/kotlin/com/amirawellness/data/models/EmotionalTrend.kt
import com.amirawellness.data.models.PeriodType // src/android/app/src/main/kotlin/com/amirawellness/data/models/EmotionalTrend.kt
import com.amirawellness.core.constants.AppConstants.EmotionType // src/android/app/src/main/kotlin/com/amirawellness/core/constants/AppConstants.kt
import com.amirawellness.core.utils.LogUtils // src/android/app/src/main/kotlin/com/amirawellness/core/utils/LogUtils.kt
import dagger.hilt.android.lifecycle.HiltViewModel // dagger.hilt:hilt-android-compiler:2.44
import kotlinx.coroutines.flow.MutableStateFlow // kotlinx.coroutines:kotlinx-coroutines-android:1.7.1
import kotlinx.coroutines.flow.StateFlow // kotlinx.coroutines:kotlinx-coroutines-android:1.7.1
import kotlinx.coroutines.flow.asStateFlow // kotlinx.coroutines:kotlinx-coroutines-android:1.7.1
import kotlinx.coroutines.launch // kotlinx.coroutines:kotlinx-coroutines-android:1.7.1
import kotlinx.coroutines.flow.catch // kotlinx.coroutines:kotlinx-coroutines-android:1.7.1
import javax.inject.Inject // javax.inject:javax.inject:1

private const val TAG = "EmotionalTrendsViewModel"

/**
 * ViewModel for the Emotional Trends screen that manages UI state and data loading
 */
@HiltViewModel
class EmotionalTrendsViewModel @Inject constructor(
    private val getEmotionalTrendsUseCase: GetEmotionalTrendsUseCase
) : ViewModel() {

    /**
     * Creates a new EmotionalTrendsViewModel with the required use case
     * @param getEmotionalTrendsUseCase
     */
    @Inject
    constructor(getEmotionalTrendsUseCase: GetEmotionalTrendsUseCase) : this(getEmotionalTrendsUseCase) {
        // Call super() to initialize ViewModel
        // Initialize getEmotionalTrendsUseCase property with the injected instance
        // Initialize _uiState with MutableStateFlow containing initial EmotionalTrendsUiState.Loading
        _uiState.value = EmotionalTrendsUiState.Loading
        // Initialize uiState as a read-only StateFlow from _uiState
        // Initialize _selectedPeriodType with MutableStateFlow containing PeriodType.WEEK as default
        _selectedPeriodType.value = PeriodType.WEEK
        // Initialize selectedPeriodType as a read-only StateFlow from _selectedPeriodType
        // Set userId to a temporary value (will be set properly when loadData is called)
        // Call loadData() to load initial data
        loadData("tempUserId")
    }

    private val _uiState = MutableStateFlow<EmotionalTrendsUiState>(EmotionalTrendsUiState.Loading)
    val uiState: StateFlow<EmotionalTrendsUiState> = _uiState.asStateFlow()

    private val _selectedPeriodType = MutableStateFlow(PeriodType.WEEK)
    val selectedPeriodType: StateFlow<PeriodType> = _selectedPeriodType.asStateFlow()

    private var userId: String = ""

    /**
     * Loads emotional trend data based on the selected period type
     * @param userId
     */
    fun loadData(userId: String) {
        // Store the provided userId in the class property
        this.userId = userId
        // Set _uiState.value to EmotionalTrendsUiState.Loading
        _uiState.value = EmotionalTrendsUiState.Loading
        // Launch a coroutine in viewModelScope
        viewModelScope.launch {
            // Get the current selected period type
            val periodType = selectedPeriodType.value
            // Call getEmotionalTrendsUseCase with userId and selectedPeriodType
            getEmotionalTrendsUseCase(userId, periodType)
                // Process the result from the use case
                .fold(
                    onSuccess = { result ->
                        // If successful, update _uiState.value to EmotionalTrendsUiState.Success with the trend data
                        _uiState.value = EmotionalTrendsUiState.Success(result.trends, result.insights)
                    },
                    onFailure = { e ->
                        // If failed, update _uiState.value to EmotionalTrendsUiState.Error with the error message
                        _uiState.value = EmotionalTrendsUiState.Error(e.message ?: "Unknown error")
                        // Log any errors that occur during data loading
                        LogUtils.logError(TAG, "Error loading emotional trend data", e)
                    }
                )
        }
    }

    /**
     * Refreshes the emotional trend data with the current settings
     */
    fun refreshData() {
        // Call loadData() with the current userId to reload data
        loadData(userId)
    }

    /**
     * Sets the selected period type and reloads data
     * @param periodType
     */
    fun setPeriodType(periodType: PeriodType) {
        // Update _selectedPeriodType.value to the provided periodType
        _selectedPeriodType.value = periodType
        // Call loadData() with the current userId to reload data with the new period type
        loadData(userId)
    }

    /**
     * Calculates the most frequent emotions from the trend data
     */
    fun getMostFrequentEmotions(): List<Pair<EmotionType, Int>> {
        // Get the current UI state
        return when (val state = _uiState.value) {
            is EmotionalTrendsUiState.Success -> {
                // If state is EmotionalTrendsUiState.Success, process the trends data
                state.trends
                    // Group trends by emotion type
                    .groupBy { it.emotionType }
                    // Count occurrences of each emotion type
                    .mapValues { (_, trends) -> trends.sumOf { it.occurrenceCount } }
                    // Sort emotions by occurrence count in descending order
                    .entries.sortedByDescending { it.value }
                    // Take the top 5 most frequent emotions
                    .take(5)
                    // Return the list of emotion type and count pairs
                    .map { Pair(it.key, it.value) }
            }
            else -> {
                // If state is not Success, return an empty list
                emptyList()
            }
        }
    }

    /**
     * Gets insights related to a specific emotion type
     * @param emotionType
     */
    fun getInsightsForEmotion(emotionType: EmotionType): List<EmotionalInsight> {
        // Get the current UI state
        return when (val state = _uiState.value) {
            is EmotionalTrendsUiState.Success -> {
                // If state is EmotionalTrendsUiState.Success, filter insights related to the specified emotion
                state.insights
                    .filter { emotionType in it.relatedEmotions }
                    // Sort insights by confidence level in descending order
                    .sortedByDescending { it.confidence }
                // Return the filtered and sorted insights
            }
            else -> {
                // If state is not Success, return an empty list
                emptyList()
            }
        }
    }

    /**
     * Gets the trend data for a specific emotion type
     * @param emotionType
     */
    fun getTrendForEmotion(emotionType: EmotionType): EmotionalTrend? {
        // Get the current UI state
        return when (val state = _uiState.value) {
            is EmotionalTrendsUiState.Success -> {
                // If state is EmotionalTrendsUiState.Success, find the trend with the specified emotion type
                state.trends.find { it.emotionType == emotionType }
                // Return the found trend or null if not found
            }
            else -> {
                // If state is not Success, return null
                null
            }
        }
    }

    /**
     * Sealed class representing the different states of the Emotional Trends UI
     */
    sealed class EmotionalTrendsUiState {
        /**
         * Default constructor for sealed class
         */
        object Loading : EmotionalTrendsUiState() {
            /**
             * Default constructor for object
             */
        }

        /**
         * Data class representing the success state with loaded emotional trend data
         * @param trends
         * @param insights
         */
        data class Success(val trends: List<EmotionalTrend>, val insights: List<EmotionalInsight>) : EmotionalTrendsUiState() {
            /**
             * Creates a Success state with the provided trend data
             * @param trends
             * @param insights
             */
            init {
                // Initialize trends property with the provided trends list
                // Initialize insights property with the provided insights list
            }
        }

        /**
         * Data class representing the error state with an error message
         * @param message
         */
        data class Error(val message: String) : EmotionalTrendsUiState() {
            /**
             * Creates an Error state with the provided error message
             * @param message
             */
            init {
                // Initialize message property with the provided error message
            }
        }
    }
}